packages/open-next/src/adapters/server-adapter.ts

```diff
@@ -164,16 +165,38 @@ function setBuildIdEnv() {
   process.env.NEXT_BUILD_ID = buildId;
 }
 
-function setNextjsPrebundledReact(rawPath: string) {
+function setNextjsPrebundledReact(
+  rawPath: string,
+  headers?: Record<string, string>
+) {
   // WORKAROUND: Set `__NEXT_PRIVATE_PREBUNDLED_REACT` to use prebundled React — https://github.com/serverless-stack/open-next#workaround-set-__next_private_prebundled_react-to-use-prebundled-react
 
-  // Get route pattern
-  const route = routesManifest.find((route) =>
-    new RegExp(route.regex).test(rawPath ?? "")
+  // Apply rewrites first, right now only used for interception
+  const rewrite = routesManifest.rewrites.find(
+    (route) =>
+      new RegExp(route.regex).test(rawPath ?? "") &&
+      route.has.length > 0 &&
+      route.has[0].type === "header" &&
+      route.has[0].key === "Next-Url" &&
+      new RegExp(route.has[0].value ?? "").test(headers?.["next-url"] ?? "")
   );
+  let route: RouteDefinition | undefined;
+
+  // Get route pattern
+  if (rewrite) {
+    route = {
+      page: rewrite.destination,
```

Comment:
If you have dynamic routes, the destination is `/:dynamic`, while the routes manifest is `/[dynamic]`, and thus `isApp` will fail and cause 404 errors.

```
rewrite.destination.replace(/:([^\/]+)/g, '[$1]')
```